File: docs/api/examples/go/kafka_client_example.go
// Package main demonstrates how to use the Takhin Kafka API with kafka-go client
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/segmentio/kafka-go"
)

const (
	brokerAddr = "localhost:9092"
	topicName  = "demo-topic"
)

func main() {
	ctx := context.Background()

	// Example 1: Create topic (via admin operations)
	if err := createTopic(ctx); err != nil {
		log.Printf("Warning: failed to create topic: %v", err)
	}

	// Example 2: Produce messages
	if err := produceMessages(ctx); err != nil {
		log.Fatalf("Failed to produce messages: %v", err)
	}

	// Example 3: Consume messages
	if err := consumeMessages(ctx); err != nil {
		log.Fatalf("Failed to consume messages: %v", err)
	}

	// Example 4: Consumer group
	if err := consumerGroupExample(ctx); err != nil {
		log.Fatalf("Failed consumer group example: %v", err)
	}

	fmt.Println("All examples completed successfully!")
}

// createTopic creates a new topic using kafka-go admin client
func createTopic(ctx context.Context) error {
	conn, err := kafka.Dial("tcp", brokerAddr)
	if err != nil {
		return fmt.Errorf("dial: %w", err)
	}
	defer conn.Close()

	controller, err := conn.Controller()
	if err != nil {
		return fmt.Errorf("get controller: %w", err)
	}

	controllerConn, err := kafka.Dial("tcp", fmt.Sprintf("%s:%d", controller.Host, controller.Port))
	if err != nil {
		return fmt.Errorf("dial controller: %w", err)
	}
	defer controllerConn.Close()

	topicConfigs := []kafka.TopicConfig{
		{
			Topic:             topicName,
			NumPartitions:     3,
			ReplicationFactor: 1,
		},
	}

	err = controllerConn.CreateTopics(topicConfigs...)
	if err != nil {
		return fmt.Errorf("create topics: %w", err)
	}

	fmt.Printf("Topic '%s' created successfully\n", topicName)
	return nil
}

// produceMessages demonstrates message production
func produceMessages(ctx context.Context) error {
	// Create a writer
	writer := kafka.NewWriter(kafka.WriterConfig{
		Brokers:      []string{brokerAddr},
		Topic:        topicName,
		Balancer:     &kafka.LeastBytes{},
		RequiredAcks: -1, // Wait for all ISR replicas
		MaxAttempts:  3,
		BatchSize:    100,
		BatchTimeout: 10 * time.Millisecond,
	})
	defer writer.Close()

	// Write messages
	messages := []kafka.Message{
		{
			Key:   []byte("user-1"),
			Value: []byte(`{"action":"login","timestamp":"2024-01-01T10:00:00Z"}`),
		},
		{
			Key:   []byte("user-2"),
			Value: []byte(`{"action":"purchase","amount":99.99}`),
		},
		{
			Key:   []byte("user-1"),
			Value: []byte(`{"action":"logout","timestamp":"2024-01-01T11:00:00Z"}`),
		},
	}

	for _, msg := range messages {
		err := writer.WriteMessages(ctx, msg)
		if err != nil {
			return fmt.Errorf("write message: %w", err)
		}
		fmt.Printf("Produced message: key=%s, value=%s\n", msg.Key, msg.Value)
	}

	fmt.Printf("Produced %d messages successfully\n", len(messages))
	return nil
}

// consumeMessages demonstrates message consumption
func consumeMessages(ctx context.Context) error {
	// Create a reader
	reader := kafka.NewReader(kafka.ReaderConfig{
		Brokers:   []string{brokerAddr},
		Topic:     topicName,
		Partition: 0,
		MinBytes:  1,
		MaxBytes:  10e6, // 10MB
		MaxWait:   500 * time.Millisecond,
	})
	defer reader.Close()

	// Read messages
	fmt.Println("Reading messages from partition 0:")
	for i := 0; i < 3; i++ {
		msg, err := reader.ReadMessage(ctx)
		if err != nil {
			return fmt.Errorf("read message: %w", err)
		}
		fmt.Printf("Consumed message: offset=%d, key=%s, value=%s\n",
			msg.Offset, msg.Key, msg.Value)
	}

	return nil
}

// consumerGroupExample demonstrates consumer group usage
func consumerGroupExample(ctx context.Context) error {
	// Create a consumer group
	groupID := "demo-consumer-group"
	reader := kafka.NewReader(kafka.ReaderConfig{
		Brokers:  []string{brokerAddr},
		GroupID:  groupID,
		Topic:    topicName,
		MinBytes: 1,
		MaxBytes: 10e6,
		MaxWait:  500 * time.Millisecond,
	})
	defer reader.Close()

	fmt.Printf("Consumer group '%s' reading messages:\n", groupID)

	// Read and commit messages
	for i := 0; i < 3; i++ {
		msg, err := reader.FetchMessage(ctx)
		if err != nil {
			return fmt.Errorf("fetch message: %w", err)
		}

		fmt.Printf("Group message: partition=%d, offset=%d, key=%s\n",
			msg.Partition, msg.Offset, msg.Key)

		// Commit the message
		if err := reader.CommitMessages(ctx, msg); err != nil {
			return fmt.Errorf("commit message: %w", err)
		}
	}

	return nil
}

// Example: List topics
func listTopics(ctx context.Context) error {
	conn, err := kafka.Dial("tcp", brokerAddr)
	if err != nil {
		return fmt.Errorf("dial: %w", err)
	}
	defer conn.Close()

	partitions, err := conn.ReadPartitions()
	if err != nil {
		return fmt.Errorf("read partitions: %w", err)
	}

	// Group partitions by topic
	topics := make(map[string][]kafka.Partition)
	for _, p := range partitions {
		topics[p.Topic] = append(topics[p.Topic], p)
	}

	fmt.Println("Available topics:")
	for topic, parts := range topics {
		fmt.Printf("  %s: %d partitions\n", topic, len(parts))
	}

	return nil
}

// Example: Get offset information
func getOffsets(ctx context.Context, topic string, partition int) error {
	conn, err := kafka.DialLeader(ctx, "tcp", brokerAddr, topic, partition)
	if err != nil {
		return fmt.Errorf("dial leader: %w", err)
	}
	defer conn.Close()

	// Get first offset
	firstOffset, err := conn.ReadFirstOffset()
	if err != nil {
		return fmt.Errorf("read first offset: %w", err)
	}

	// Get last offset
	lastOffset, err := conn.ReadLastOffset()
	if err != nil {
		return fmt.Errorf("read last offset: %w", err)
	}

	fmt.Printf("Topic '%s' partition %d:\n", topic, partition)
	fmt.Printf("  First offset: %d\n", firstOffset)
	fmt.Printf("  Last offset: %d\n", lastOffset)
	fmt.Printf("  Messages: %d\n", lastOffset-firstOffset)

	return nil
}

// Example: Transactional producer
func transactionalProduce(ctx context.Context) error {
	writer := kafka.NewWriter(kafka.WriterConfig{
		Brokers:      []string{brokerAddr},
		Topic:        topicName,
		Balancer:     &kafka.LeastBytes{},
		RequiredAcks: -1,
		// Enable idempotence and transactions
		Idempotent: true,
	})
	defer writer.Close()

	// Begin transaction
	txn := writer.BeginTransaction()

	messages := []kafka.Message{
		{Key: []byte("txn-1"), Value: []byte("message 1")},
		{Key: []byte("txn-2"), Value: []byte("message 2")},
	}

	// Write messages in transaction
	if err := txn.WriteMessages(ctx, messages...); err != nil {
		txn.Abort()
		return fmt.Errorf("write messages: %w", err)
	}

	// Commit transaction
	if err := txn.Commit(); err != nil {
		return fmt.Errorf("commit transaction: %w", err)
	}

	fmt.Println("Transaction committed successfully")
	return nil
}

File: docs/api/examples/python/kafka_client_example.py
"""
Takhin Kafka Protocol Client - Python Example

This example demonstrates how to interact with Takhin using the kafka-python library.

Install dependencies:
    pip install kafka-python
"""

from kafka import KafkaProducer, KafkaConsumer, KafkaAdminClient
from kafka.admin import NewTopic, ConfigResource, ConfigResourceType
from kafka.errors import TopicAlreadyExistsError
import json
import time


BOOTSTRAP_SERVERS = ['localhost:9092']
TOPIC_NAME = 'demo-topic'


def create_topic():
    """Create a topic using Kafka Admin API"""
    admin_client = KafkaAdminClient(bootstrap_servers=BOOTSTRAP_SERVERS)
    
    topic = NewTopic(
        name=TOPIC_NAME,
        num_partitions=3,
        replication_factor=1
    )
    
    try:
        admin_client.create_topics([topic])
        print(f"Topic '{TOPIC_NAME}' created successfully")
    except TopicAlreadyExistsError:
        print(f"Topic '{TOPIC_NAME}' already exists")
    finally:
        admin_client.close()


def produce_messages():
    """Produce messages to a topic"""
    producer = KafkaProducer(
        bootstrap_servers=BOOTSTRAP_SERVERS,
        key_serializer=lambda k: k.encode('utf-8') if k else None,
        value_serializer=lambda v: json.dumps(v).encode('utf-8'),
        acks='all',  # Wait for all ISR replicas
        retries=3
    )
    
    messages = [
        {"key": "user-1", "value": {"action": "login", "timestamp": "2024-01-01T10:00:00Z"}},
        {"key": "user-2", "value": {"action": "purchase", "amount": 99.99}},
        {"key": "user-1", "value": {"action": "logout", "timestamp": "2024-01-01T11:00:00Z"}},
    ]
    
    print("\n=== Producing Messages ===")
    for msg in messages:
        future = producer.send(
            TOPIC_NAME,
            key=msg["key"],
            value=msg["value"],
            partition=0
        )
        
        # Wait for send to complete
        record_metadata = future.get(timeout=10)
        print(f"Produced: key={msg['key']}, offset={record_metadata.offset}, "
              f"partition={record_metadata.partition}")
    
    producer.flush()
    producer.close()


def consume_messages():
    """Consume messages from a topic"""
    consumer = KafkaConsumer(
        TOPIC_NAME,
        bootstrap_servers=BOOTSTRAP_SERVERS,
        auto_offset_reset='earliest',
        enable_auto_commit=False,
        consumer_timeout_ms=5000,  # Stop after 5 seconds of no messages
        value_deserializer=lambda v: json.loads(v.decode('utf-8'))
    )
    
    print("\n=== Consuming Messages ===")
    for message in consumer:
        print(f"Consumed: partition={message.partition}, offset={message.offset}, "
              f"key={message.key.decode('utf-8') if message.key else None}, "
              f"value={message.value}")
    
    consumer.close()


def consumer_group_example():
    """Example using consumer groups"""
    GROUP_ID = 'demo-consumer-group'
    
    consumer = KafkaConsumer(
        TOPIC_NAME,
        bootstrap_servers=BOOTSTRAP_SERVERS,
        group_id=GROUP_ID,
        auto_offset_reset='earliest',
        enable_auto_commit=True,
        auto_commit_interval_ms=1000,
        value_deserializer=lambda v: json.loads(v.decode('utf-8'))
    )
    
    print(f"\n=== Consumer Group '{GROUP_ID}' ===")
    count = 0
    for message in consumer:
        print(f"Group message: partition={message.partition}, offset={message.offset}, "
              f"key={message.key.decode('utf-8') if message.key else None}")
        count += 1
        if count >= 3:
            break
    
    consumer.close()


def describe_topic():
    """Describe topic configuration"""
    admin_client = KafkaAdminClient(bootstrap_servers=BOOTSTRAP_SERVERS)
    
    # Describe topic configs
    resource = ConfigResource(ConfigResourceType.TOPIC, TOPIC_NAME)
    configs = admin_client.describe_configs([resource])
    
    print(f"\n=== Topic Configuration: {TOPIC_NAME} ===")
    for config_resource, config_entries in configs.items():
        for config_key, config_entry in config_entries.items():
            print(f"  {config_key} = {config_entry.value}")
    
    admin_client.close()


def list_consumer_groups():
    """List all consumer groups"""
    admin_client = KafkaAdminClient(bootstrap_servers=BOOTSTRAP_SERVERS)
    
    groups = admin_client.list_consumer_groups()
    
    print("\n=== Consumer Groups ===")
    for group_id, group_type in groups:
        print(f"  Group: {group_id}, Type: {group_type}")
    
    admin_client.close()


def get_offsets():
    """Get partition offsets"""
    consumer = KafkaConsumer(
        bootstrap_servers=BOOTSTRAP_SERVERS,
        enable_auto_commit=False
    )
    
    # Get partitions for topic
    partitions = consumer.partitions_for_topic(TOPIC_NAME)
    
    print(f"\n=== Partition Offsets for {TOPIC_NAME} ===")
    for partition in sorted(partitions):
        # Get beginning offset
        tp = {'topic': TOPIC_NAME, 'partition': partition}
        consumer.assign([tp])
        consumer.seek_to_beginning(tp)
        beginning = consumer.position(tp)
        
        # Get end offset
        consumer.seek_to_end(tp)
        end = consumer.position(tp)
        
        print(f"  Partition {partition}: beginning={beginning}, end={end}, messages={end-beginning}")
    
    consumer.close()


def transactional_produce():
    """Example of transactional producer"""
    producer = KafkaProducer(
        bootstrap_servers=BOOTSTRAP_SERVERS,
        transactional_id='my-transactional-producer',
        key_serializer=lambda k: k.encode('utf-8') if k else None,
        value_serializer=lambda v: json.dumps(v).encode('utf-8'),
        acks='all',
        enable_idempotence=True
    )
    
    producer.init_transactions()
    
    print("\n=== Transactional Produce ===")
    try:
        producer.begin_transaction()
        
        messages = [
            {"key": "txn-1", "value": {"data": "message 1"}},
            {"key": "txn-2", "value": {"data": "message 2"}},
        ]
        
        for msg in messages:
            producer.send(TOPIC_NAME, key=msg["key"], value=msg["value"])
            print(f"Sent: key={msg['key']}")
        
        producer.commit_transaction()
        print("Transaction committed successfully")
    except Exception as e:
        print(f"Transaction failed: {e}")
        producer.abort_transaction()
    finally:
        producer.close()


def main():
    """Run all examples"""
    print("Takhin Kafka Client Examples\n")
    
    # Create topic
    create_topic()
    time.sleep(1)
    
    # Produce messages
    produce_messages()
    time.sleep(1)
    
    # Consume messages
    consume_messages()
    time.sleep(1)
    
    # Consumer group example
    consumer_group_example()
    time.sleep(1)
    
    # Describe topic
    describe_topic()
    
    # List consumer groups
    list_consumer_groups()
    
    # Get offsets
    get_offsets()
    
    # Transactional produce
    transactional_produce()
    
    print("\nAll examples completed!")


if __name__ == "__main__":
    main()
